Support dark mode

dscho opened the pull request:
## Changes

This adds support for dark mode.

## Context

- #1432 

It went through quite a few iterations, with contributions by @rimrul, @TheGiraffe3, and @To1ne, as well as a few other people who provided feedback to me privately.

dscho commented:
This will need _a ton_ of work.

There are way too many hard-coded colors in the current version of our CSS. They will need to be expressed in terms of CSS custom properties, and I have a hunch that there lurks also a really good opportunity to use not _quite_ as many different colors, but to reduce by reusing fewer colors, which should also make the overall appearance nicer.

However, I think I saw a couple of color definitions that were not even used. So before trying to encapsulate the color palette, we should probably use something like https://github.com/purifycss/purifycss to identify parts that are no longer needed (I manually identified [three lines already](https://github.com/git/git-scm.com/pull/1888/commits/422ea88b5c669034f154a15e446d045165bbfa9f) that can easily go and stop bothering us). No need to adapt unused CSS.

And finally, as I said elsewhere, I have no illusions about being a designer, and hope that someone with a proven record of a better design taste than myself (@To1ne maybe?) will give this some TLC.

dscho commented:
For shiggles, I branch-deployed this to https://dscho.github.io/git-scm.com/.

rimrul commented:
> For shiggles, I branch-deployed this to https://dscho.github.io/git-scm.com/.

Looking at that on Firefox 130.0.1 on Android it has some issues still.

The dark mode just ends appruptly.

![Screenshot_20240928_082946_Firefox](https://github.com/user-attachments/assets/11294a7a-9295-41bf-93ef-50389a6c2fda)

Switching to light mode only toggles the text, but not the background.

![Screenshot_20240928_083019_Firefox](https://github.com/user-attachments/assets/9da4679c-76a4-4703-b3d8-48841b19b9ad)

rimrul commented:
But I do think the parts that work look quite nice.

dscho commented:
> The dark mode just ends appruptly.

Right, that's one thing I could not figure out. You will notice [the `height: 200%`](https://github.com/git/git-scm.com/pull/1888/files#diff-d692267383cf5a41e7e464c75709ab25035c90465578b0a49bd26dd978b68d88R31) thing I tried, but it only worked around it on a few pages (and only on Desktop). I guess the real fix will be to have it `repeat`, just like [the isometric grid](https://github.com/git/git-scm.com/pull/1888/files#diff-d692267383cf5a41e7e464c75709ab25035c90465578b0a49bd26dd978b68d88R44) on the front page.

> But I do think the parts that work look quite nice.

Thank you!

There's still a ton of work to do, in particular cleaning up the CSS and then consolidating the remaining hard-coded colors into CSS custom properties à la `--main-bg`. Want to join the fun?

dscho commented:
I've decided to punt on the CSS clean-up for now ;-)

@rimrul @To1ne I've updated this branch and deployed it to https://dscho.github.io/git-scm.com/; Would you mind giving it another go?

For the record, here are a couple of Firefox Mobile screenshots:
<details><summary>Front page</summary>

![image](https://github.com/user-attachments/assets/6187969c-3bd7-4b34-b32d-0ca2ac169a74)
</details>
<details><summary>About - Branching and Merging</summary>

![image](https://github.com/user-attachments/assets/080735c9-dbdd-4d16-80f4-35588f788085)
</details>
<details><summary>Sidebar</summary>

![image](https://github.com/user-attachments/assets/cc0bf5bd-c33a-4d6f-a76f-ce732670aca6)
</details>
<details><summary>Documentation</summary>

![image](https://github.com/user-attachments/assets/36c92a42-c361-42d7-91af-5ebb4625e799)
</details>
<details><summary>A manual page</summary>

![image](https://github.com/user-attachments/assets/a7492245-c701-4662-b986-1ece2472f098)
</details>
<details><summary>Versions drop-down of a manual page</summary>

![image](https://github.com/user-attachments/assets/0d626dd3-9614-4e94-9896-fcb041bdc516)
</details>
<details><summary>Book, chapter 1 section 1</summary>

![image](https://github.com/user-attachments/assets/4bf58939-4b66-4306-a7ab-6cd96f9589ca)
</details>
<details><summary>Downloads</summary>

Here, I present both the dark mode and the light mode to demonstrate that the monitor is cut off in both, therefore that's a bug that was not introduced by this PR:

| Dark Mode | Light mode |
| - | - |
| ![image](https://github.com/user-attachments/assets/eeae9f77-80ce-43c1-a2ac-3bde522729ab) | ![image](https://github.com/user-attachments/assets/92c1d8ce-ca32-4aa9-be29-e5464713e6f8) |
</details>

To1ne commented:
@dscho I've had a more thorough look at it. Overall I'm quite impressed!

Shout outs:

* It's nice to see it works without Javascript.
* It's also very nice it remembers your preference (when JS is enabled). Although I don't know if you can reset this state to follow the browser/OS again?
* I like the overall aesthetic. Colors feels comfy and familiar and contrast looks decent.

Few things:

* The contrast of the Git logo is too low. Text should be white(ish).
* Same about the logo at the bottom.
* We should remove the background from the brand logos at the bottom of the landing page.
* I'm not sure I like dashed underline on links. But this could become our aesthetic, but should light mode have it as well?
* The color of the underline looks weird in some cases, when text color is different from underline, like here:
  ![image](https://github.com/user-attachments/assets/9c698cda-512a-46a1-9980-3cad0670ec3c)
* Search menu looks bad with underline hyperlinks:
  ![20241106_07h02m02s_grim](https://github.com/user-attachments/assets/9ffb1da7-2c81-4f0e-9094-e198682fd28b)
* I'm not a fan of the sun/moon animation (when you load the page)
* I think I would move the sun/moon to the top right of the screen, it seems to be more common done like this. And then it doesn't clash with the "scroll to top" button. You could even move it right next to the search box.
* Some of our images on `/about/*` should be updated to work better with dark mode. But that can be done in a later iteration. These images might need an overhaul anyway.
* Border on `<code>` feels too heavy:
  ![image](https://github.com/user-attachments/assets/6c904386-5229-4153-9baf-2c698d9ab9f7)
* I'm not sure we should apply the brightness filter to all images, like for example on `/downloads/logos` or `/about/data-assurance`
* The color of `<span class="">` on `/community` looks, well, not good:
![image](https://github.com/user-attachments/assets/39b02ad6-01e9-4694-a36d-70f20ccc124e)
* I'm not a fan of the aesthetic of the sun & moon. I think monochrome icons would fit in better, but it's cool it just works with emoji, so I get it.

All this is based on visual review, I didn't review any of the code. I can, if you like to?

If you like I can submit some patches with suggestions?

dscho commented:
> If you like I can submit some patches with suggestions?

I would love that!

To1ne commented:
> > If you like I can submit some patches with suggestions?
> I would love that!

Maybe next week.

TheGiraffe3 commented:
Just one problem that I found: some icons still have a bright background.

![Screen Shot 2024-11-28 at 8 47 10 AM](https://github.com/user-attachments/assets/8ad8cf1f-f389-4af1-8990-ccd0413acff1)

Other than that, wonderful! This will be quite nice to have.

dscho commented:
@TheGiraffe3 could you help out and contribute fixed icons, preferably as a PR to my branch?

TheGiraffe3 commented:
I'll try.

TheGiraffe3 commented:
There's a pull request up for that: https://github.com/dscho/git-scm.com/pull/16.

To1ne commented:
@TheGiraffe3 Cool work. Do you happen to have screenshots of both in light and dark mode?

@dscho About the logos in general. I think we should reconsider which/if they are relevant, but that's something for a separate MR/issue. Shall we open an issue on that? Or ask on the mailing list?

My idea was to simply put the light background on all logos in CSS, even in dark mode (so the images without transparent background wouldn't stand out against the rest), but I haven't gotten to that yet. Then when we know which logos we want to keep/add/remove we could update the images accordingly. But thanks to @TheGiraffe3 that's not needed no more.

TheGiraffe3 commented:
> @TheGiraffe3 Cool work. Do you happen to have screenshots of both in light and dark mode?

On the website?
I'm afraid I haven't yet been able to figure out how to deploy it with the changed icons, so no, not at the moment.

To1ne commented:
> On the website?

@TheGiraffe3 doesn't need to be a deployed version, can be a screenshot when running the site locally. I suggest you have a look at the README and let us know if anything isn't clear how to preview the site locally?

Anyhow, I've taken the liberty of running your changes locally:

Light:

![image](https://github.com/user-attachments/assets/f3df673b-4cbd-4415-9fe0-22c0e7080384)

Dark:

![image](https://github.com/user-attachments/assets/4137ef10-a544-4658-aaa2-a83e14fd83bc)

Just as reference, this is the current original:

![image](https://github.com/user-attachments/assets/1527e622-2b46-49a2-b41c-300b6c0fe167)

And this is the dark version @dscho 's branch:

![image](https://github.com/user-attachments/assets/0f4ded25-a12a-4cc9-a7be-950665ad3fcb)

TheGiraffe3 commented:
I'll just note here that the old Eclipse logo was out of date by two versions and the background proved hard to remove, so we switched the logo out when changing it anyway.

dscho commented:
> the old Eclipse logo was out of date by two versions

I should probably clarify that the "middle version" was an April Fools' joke:

![image](https://github.com/user-attachments/assets/f9523198-f8cf-44ba-bec9-332bfb76e8a1)

In any case, here is the comparison after merging @TheGiraffe3's excellent work (thank you so much!):

Light mode:

![image](https://github.com/user-attachments/assets/2291195a-09fe-4543-befa-a2594a93b7d6)

Dark mode:

![image](https://github.com/user-attachments/assets/2bc21c3b-6ca9-4345-95bb-2e4c6a7d52ad)

rimrul commented:
The Microsoft logo in dark mode has been mangled quite a bit

dscho commented:
> The Microsoft logo in dark mode has been mangled quite a bit

Hmm. I'm not even sure that this is a legitimate use of the trademark, see https://www.microsoft.com/en-us/legal/intellectualproperty/Trademarks/...

dscho commented:
> > The Microsoft logo in dark mode has been mangled quite a bit
> 
> Hmm. I'm not even sure that this is a legitimate use of the trademark, see https://www.microsoft.com/en-us/legal/intellectualproperty/Trademarks/...

To be precise, I am concerned about this guideline: _Don’t imply an affiliation, endorsement, sponsorship, or approval with or by Microsoft._

dscho commented:
@To1ne I _think_ that I have addressed all of the concerns you raised in https://github.com/git/git-scm.com/pull/1888#issuecomment-2458954252, could I ask you to take another look? As usual, I've test-deployed the changes to https://dscho.github.io/git-scm.com/.

To1ne commented:
So, what's left?

dscho commented:
> So, what's left?

I think that's it 😉 

Should I merge it now?

To1ne commented:
@dscho Do you know how to remove the `text-shadow` from the `a.highlight` and `a:hover` in the search dropdown?

Also I've found one quirk, when you manually set it to light mode, but your system says dark mode, you get this.

![Screen Shot 2025-03-20 at 07 45 25](https://github.com/user-attachments/assets/38312f59-8971-4a4a-a2fb-e1a55d960cda)

dscho commented:
> Also I've found one quirk, when you manually set it to light mode, but your system says dark mode, you get this.

Hah, I forgot about this. The problem here is that I designed the "dark mode" settings to override the light-mode settings, but in dark mode, it became the default and the light-mode toggle could not override that. I fixed that like this:

```diff
diff --git a/assets/sass/dark-mode.css b/assets/sass/dark-mode.css
index af36be6cc..fc51f2d76 100644
--- a/assets/sass/dark-mode.css
+++ b/assets/sass/dark-mode.css
@@ -184,6 +184,6 @@
 @include mode($mode: dark, $theme: '[data-theme="dark"]')
 
 @media screen and (prefers-color-scheme: dark) {
-    @include mode($mode: dark)
-    @include mode($mode: light, $theme: '[data-theme="light"]')
+    @include mode($mode: dark, $theme: ':not([data-theme="light"])')
+    @include mode($mode: light)
 }
```

> Do you know how to remove the `text-shadow` from the `a.highlight` and `a:hover` in the search dropdown?

Done: 9675370f7a2ab4534a78c01af2e062ebbdd03c63

To1ne commented on tests/git-scm.spec.js:
```diff
@@ -290,3 +290,32 @@ test('small-and-fast', async ({ page }) => {
+
+  await page.goto(`${url}`);
+  await page.evaluate(() => {
+    document.querySelector('#tagline').innerHTML = '--dark-mode-for-dark-times';
```
LOL!

dscho commented:
@To1ne do you think this is ready to merge? I kind of want to, and then take it from there.

To1ne commented:
> do you think this is ready to merge? I kind of want to, and then take it from there.

@dscho I think we are. I'm kind of tired working in a fork on this.

One question though: Do we have/need a rollback strategy if we learn these changes were not ready to be published? I don't think it's needed, like what's the worst that can happen? And we can always do `git revert` if needed.

dscho commented:
> Do we have/need a rollback strategy if we learn these changes were not ready to be published?

Yes, there is a "Revert" button once this PR is merged. And with that, I'll merge!
